source/hpack_decoder.c

```diff
@@ -278,6 +278,25 @@ int aws_hpack_decode(
                     decoder->progress_entry.u.literal.prefix_size = 4;
                     decoder->progress_entry.state = HPACK_ENTRY_STATE_LITERAL_BEGIN;
                 }
+                if (decoder->dynamic_table_protocol_max_size_setting < decoder->context.dynamic_table.size) {
```

Comment:
I don't think the logic here is correct

The encoder has the correct logic. In the `dynamic_table_size_update` struct it tracks `latest_value` and `smallest_value` separately.
```
    struct {
        size_t latest_value;
        size_t smallest_value;
        bool pending;
    } dynamic_table_size_update;
```

The decoder needs to do something similar, if multiple SETTINGS packets are received back to back, and change the table size to 10, then to 100, then the decoder needs to confirm that it receives an update `<= 10` (RFC-7541 4.2)

Reply:
Hmmm, I read again, and still feel confusing about:
> An endpoint MUST treat a field block that follows an acknowledgment of the reduction to the maximum dynamic table size as a [connection error](https://httpwg.org/http2-spec/draft-ietf-httpbis-http2bis.html#ConnectionErrorHandler) ([Section 5.4.1](https://httpwg.org/http2-spec/draft-ietf-httpbis-http2bis.html#ConnectionErrorHandler)) of type [COMPRESSION_ERROR](https://httpwg.org/http2-spec/draft-ietf-httpbis-http2bis.html#COMPRESSION_ERROR) if it does not start with a conformant Dynamic Table Size Update instruction.

[here](https://httpwg.org/http2-spec/draft-ietf-httpbis-http2bis.html#section-4.3.1-4) should be the requirement from the decoder. Especially about `start with a conformant Dynamic Table Size Update instruction`. *What is conformant update instruction for the case you pointing out to start with?* I don't think there is any clear statement about what kind of `Dynamic Table Size Update instruction` to start with.

from hpack spec [here](https://www.rfc-editor.org/rfc/rfc7541#section-4.2):
>the smallest maximum table size that occurs in that interval MUST be signaled in a dynamic table
   size update.  The final maximum size is always signaled, resulting in at most two dynamic table size updates

I guess the first one could be larger than the smallest value as long as the second one is smaller than it. :(

Reply:
> resulting in **at most two** dynamic table size updates

However, I don't see any place that enforce the dynamic table size updates to be no more than one in normal case. But I guess they assume that's the case as from [here](https://www.rfc-editor.org/rfc/rfc7541#section-6.3)
>In HTTP/2, this limit is **the last value** of the SETTINGS_HEADER_TABLE_SIZE parameter

Here is my understanding:

- We can receive at most two continuous Dynamic Table Size Update instruction after ack the settings.
- One of those two MUST be smaller than the smallest max.
- Everyone MUST be smaller than the latest max.